text/0000-http-index.md

```diff
@@ -0,0 +1,139 @@
+
+# Motivation
+[motivation]: #motivation
+
+The full crate index is relatively big and slow to download. It will keep growing as crates.io grows, making the problem worse. The requirement to download the full index slows down the first use of Cargo. It's especially slow and wasteful in stateless CI environments, which download the full index, use only a tiny fraction of it, and throw it away. Caching of the index in hosted CI environments is difficult (`.cargo` dir is large) and often not effective (e.g. upload and download of large caches in Travis CI is almost as slow as a fresh index download).
+
+The kind of data stored in the index is not a good fit for the git protocol. The index content (as of eb037b4863) takes 176MiB as an uncompressed tarball, 16MiB with `gz -1`, and 10MiB compressed with `xz -6`. Git clone reports downloading 215MiB. That's more than just the uncompressed latest index content, and over **20 times more** than a compressed tarball.
+
+A while ago, GitHub indicated they [don't want to support shallow clones of large repositories](http://blog.cocoapods.org/Master-Spec-Repo-Rate-Limiting-Post-Mortem/). libgit2 doesn't support shallow clones yet. Squashing of the index history adds complexity to management and consumption of the index (which is also used by tools other than Cargo), and still doesn't solve problems of the git protocol inefficiency and overall growth.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Expose the index over HTTP as simple files, keeping the existing content and directory layout unchanged (the existing raw.githubusercontent.com view may even be enough for this). The current format is structured like this:
+
+```
+/config.json
+/ac/ti
+/ac/ti/action
+/ac/ti/actiondb
+/ac/ti/actions
+/ac/ti/actions-toolkit-sys
+/ac/ti/activation
+/ac/ti/activeds-sys
+…
+```
+
+To learn about crates and resolve dependencies, Cargo (or any other client) would make requests to known URLs for each dependency it needs to learn about, e.g. `https://index.example.com/se/rd/serde`. For each dependency the client would also have to request information about its dependencies, recursively, until all dependencies are fetched (and cached) locally.
+
+It's possible to request dependency files in parallel, so the worst-case latency of such dependency resolution is limited to the maximum depth of the dependency tree. In practice it's less, because dependencies occur in multiple places in the tree, allowing earlier discovery and increasing parallelization. Additionally, if there's a lock file, all dependencies listed in it can be speculatively checked in parallel. Similarly, cached dependency files can be used to speculatively check known sub-dependencies sooner.
```

Comment:
Perhaps counter-intuitively, that last optimization (using cached dependency files while we're still downloading the updated one) ends up _slowing down_ the index process, if only by a little:
![cdn](https://user-images.githubusercontent.com/176295/101223935-126ff280-3642-11eb-940c-a94d7743f3c5.png)
I _believe_ this is because the optimization doesn't end up speeding up fetches by too much (there is already enough fetch parallelism), but it _does_ add a computational cost in terms of a second iteration through the same dependency graph, including parsing each double-checked index file a second time, which ends up adding more than it brings to the table.

Reply:
That's interesting. I assumed that cost of all local operations is essentially zero compared to the network cost.

Could it be that Cargo adds some unexpected overhead that's not as simple as parsing JSON and matching semver?

Reply:
Yeah, I was surprised too. I think the overhead comes not only from running the parser+matching twice, but from having to do the whole greedy fetch walk twice. Even though the second walk is unlikely to lead to many downloads, we still need to go through the logic to walk the dependency graph. And that adds up it seems like.